rpm/packagescanner.go

```diff
@@ -0,0 +1,296 @@
+	ctx := context.TODO()
+	ctx = log.Logger.WithContext(ctx)
+	return ps.ScanContext(ctx, layer)
+}
+
+// ScanContext attempts to find rpm databases within the layer and enumerate the
+// packages there.
+//
+// A return of (nil, nil) is expected if there's no rpm database.
+//
+// The external commands "tar" and "rpm" are used and expected to be in PATH.
+func (ps *Scanner) ScanContext(ctx context.Context, layer *claircore.Layer) ([]*claircore.Package, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	trace.Log(ctx, "layer:sha256", layer.Hash)
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "package_scanner").
+		Str("name", ps.Name()).
+		Str("version", ps.Version()).
+		Str("kind", ps.Kind()).
+		Str("layer", layer.Hash).
+		Logger()
+	log.Debug().Msg("start")
+	defer log.Debug().Msg("done")
+
+	r, err := layer.Reader()
+	if err != nil {
+		return nil, err
+	}
+	rd, ok := r.(interface {
+		io.ReadCloser
+		io.Seeker
+	})
+	if !ok {
+		return nil, errors.New("")
+	}
+
+	possible := make(map[string]int)
+	tr := tar.NewReader(rd)
+	// Find possible rpm dbs
+	// If none found, return
+	var h *tar.Header
+	for h, err = tr.Next(); err == nil; h, err = tr.Next() {
+		n := filepath.Base(h.Name)
+		d := filepath.Dir(h.Name)
+		if _, ok := dbnames[n]; ok {
+			possible[d]++
+		}
+	}
+	if err != io.EOF {
+		return nil, err
+	}
+	if len(possible) == 0 {
+		return nil, nil
+	}
+	found := make([]string, 0)
+	for k, score := range possible {
+		if score == len(dbnames) {
+			found = append(found, k)
+		}
+	}
+	log.Debug().Int("count", len(found)).Msg("found possible databases")
+	if len(found) == 0 {
+		return nil, nil
+	}
+
+	root, err := ioutil.TempDir("", "rpmscanner.")
+	if err != nil {
+		return nil, err
+	}
+	// Extract tarball
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	if _, err := rd.Seek(0, io.SeekStart); err != nil {
+		return nil, fmt.Errorf("rpm: unable to seek: %w", err)
+	}
+	tarcmd := exec.CommandContext(ctx, "tar", "-xC", root)
+	tarcmd.Stdin = rd
+	if err := tarcmd.Run(); err != nil {
+		return nil, fmt.Errorf("rpm: failed to untar: %w", err)
+	}
+	log.Debug().Str("dir", root).Msg("extracted layer")
+	defer func() {
+		err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+			// If a directory isn't o+w, fix it.
+			if m := info.Mode(); info.IsDir() && m&0200 == 0 {
+				return os.Chmod(path, m|0200)
+			}
+			return nil
+		})
+		if err != nil {
+			log.Warn().Err(err).Msg("error removing extracted files")
+		}
+		if err := os.RemoveAll(root); err != nil {
+			log.Warn().Err(err).Msg("error removing extracted files")
+		}
+	}()
+
+	var pkgs []*claircore.Package
+	// Using --root and --dbpath, run rpm query on every suspected database
+	for _, db := range found {
+		log.Debug().Str("db", db).Msg("examining database")
+		r, w, err := os.Pipe()
```

Comment:
Why not drop the pipe here and use the stderr and stdout pipes the "*exec.CMD" struct provides? You will then be able to check and log errors coming from the binary invocation which is missing here.

https://golang.org/src/os/exec/exec.go?s=18556:18605#L628